ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/server/BackupReplayFileTest.java

```diff
@@ -0,0 +1,173 @@
+
+import io.confluent.ksql.rest.entity.CommandId;
+import io.confluent.ksql.rest.server.computation.Command;
+import io.confluent.ksql.util.Pair;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+
+@RunWith(MockitoJUnitRunner.class)
+public class BackupReplayFileTest {
+  private static final String KEY_VALUE_SEPARATOR = ":";
+  private static final String REPLAY_FILE_NAME = "backup_command_topic_1";
+
+  @Rule
+  public TemporaryFolder backupLocation = new TemporaryFolder();
+
+  private BackupReplayFile replayFile;
+  private File internalReplayFile;
+
+  @Before
+  public void setup() throws IOException {
+    internalReplayFile = backupLocation.newFile(REPLAY_FILE_NAME);
+    replayFile = new BackupReplayFile(internalReplayFile);
+  }
+
+  @Test
+  public void shouldGetFilePath() {
+    // When
+    final String path = replayFile.getPath();
+
+    // Then
+    assertThat(path, is(String.format(
+        "%s/%s", backupLocation.getRoot().getAbsolutePath(), REPLAY_FILE_NAME)));
+  }
+
+  @Test
+  public void shouldWriteRecord() throws IOException {
+    // Given
+    final Pair<CommandId, Command> record = newStreamRecord("stream1");
+
+    // When
+    replayFile.write(record.left, record.right);
+
+    // Then
+    final List<String> commands = Files.readAllLines(internalReplayFile.toPath());
+    assertThat(commands.size(), is(1));
+    assertThat(commands.get(0), is(
+        "\"stream/stream1/create\"" + KEY_VALUE_SEPARATOR
+            + "{\"statement\":\"CREATE STREAM stream1 (id INT) WITH (kafka_topic='stream1')\"}"
+    ));
+  }
+
+  @Test
+  public void shouldWriteListOfRecords() throws IOException {
+    // Given
+    final Pair<CommandId, Command> record1 = newStreamRecord("stream1");
+    final Pair<CommandId, Command> record2 = newStreamRecord("stream2");
+
+    // When
+    replayFile.write(Arrays.asList(record1, record2));
+
+    // Then
+    final List<String> commands = Files.readAllLines(internalReplayFile.toPath());
+    assertThat(commands.size(), is(2));
+    assertThat(commands.get(0), is(
+        "\"stream/stream1/create\"" + KEY_VALUE_SEPARATOR
+            + "{\"statement\":\"CREATE STREAM stream1 (id INT) WITH (kafka_topic='stream1')\"}"
+    ));
+    assertThat(commands.get(1), is(
+        "\"stream/stream2/create\"" + KEY_VALUE_SEPARATOR
+            + "{\"statement\":\"CREATE STREAM stream2 (id INT) WITH (kafka_topic='stream2')\"}"
+    ));
+  }
+
+  @Test
+  public void shouldWriteRecordWithNewLineCharacterInCommand() throws IOException {
```

Comment:
I think my concern was more about reading record with new line characters to make sure that `File.readLines` works properly

Reply:
The only way to make it fail is to add a `\n` directly in the file, like this:
```
final String commandWithNewLineChar = "\"stream/stream1/create\"" + KEY_VALUE_SEPARATOR
        + "{\"statement\":\"CREATE STREAM stream1 (id\n INT) WITH (kafka_topic='stream1')\"}";

Files.write(internalReplayFile.toPath(),
        commandWithNewLineChar.getBytes(StandardCharsets.UTF_8));
```

`File.readLines` will read the line up to the `id\n`. The deserializer fails because the command line is not valid. In these cases, the `CommandTopicBackup` will catch the error, log it, and create a new backup file as this current one is corrupted.

I don't think we'll see a scenario with the `\n` character like that. Say KSQL supports the `\n`, then the backup will write the `\n` like the `shouldWriteRecordWithNewLineCharacterInCommand`, and `readLines` will work fine. But if a user goes directly to the file and modifies the `\n`, and the file gets corrupted, which is expected to fail.